Add the ability to merge the raid hp bar overlay with the opponent info overlay

CasvM opened the pull request:
## example:

### before:
![image](https://user-images.githubusercontent.com/20537001/94916413-ef448d80-04ae-11eb-8ae0-859cd092b0c2.png)

### after: 
![image](https://user-images.githubusercontent.com/20537001/94916490-0d11f280-04af-11eb-9bef-a008b9eb8b6a.png)


This would essentially close these isssues: https://github.com/runelite/runelite/issues/12283 https://github.com/runelite/runelite/issues/12230 https://github.com/runelite/runelite/issues/12137

Hydrox6 commented:
How does this work with Vanguards, considering they have a really overengineered health bar that tries to show all 3 at once iirc (#12283 would want vanilla for *just* vanguards, likely because of this)

Hydrox6 reviewed (changes requested):
You've got a lot of checkstyle violations. You can use this page to get our style guide and import our formatter

https://github.com/runelite/runelite/wiki/Code-Conventions

CasvM commented:
I've added the ability to force the vanilla HP bar by typing it in the config:
![image](https://user-images.githubusercontent.com/20537001/94966291-4ec98a00-04fd-11eb-88c0-b3a64b4d2fe8.png)

I've also added the percentage to the vanilla HP bar 
![image](https://user-images.githubusercontent.com/20537001/94966935-67866f80-04fe-11eb-96d2-80b60b3ab6e2.png)

I'm not sure why the checks are failing, I formatted the code before submitting.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoPlugin.java:
```diff
@@ -164,6 +165,43 @@ public void onGameTick(GameTick gameTick)
+	@Subscribe
+	public void onConfigChanged(ConfigChanged event)
+	{
+		if (event.getGroup().equals("opponentinfo") || event.getGroup().equals("runelite"))
```
Why is `runelite` in here? I don't really understand this at all as written.

CasvM replied:
When the plugin itself is enabled or disabled through the config it should also remove/unhide the health bar, but I forgot I could do that through the startup and shutdown. I've removed the runelite part in the latest commit

Nightfirecat commented on runelite-api/src/main/java/net/runelite/api/Varbits.java:
```diff
@@ -304,6 +304,8 @@
 	PERSONAL_POINTS(5422),
 	RAID_PARTY_SIZE(5424),
 
+	RAID_HEALTH_BAR_OVERLAY_VALUE(6099),
```
What value does this hold? Seems like it's an int of the current health of the monster? Add a comment to describe this.

CasvM replied:
I've added another varbit for the max health of a monster, and a comment to describe these varbits

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoConfig.java:
```diff
@@ -63,4 +63,39 @@ default boolean showOpponentsInMenu()
+		position = 4
+	)
+
+	default boolean showVanillaPercentages()
```
Isn't this already covered by `HitpointsDisplayStyle.BOTH`?

CasvM replied:
While I could make it so that the value on the vanilla health bar is equal to that of the opponent information, I thought someone might want just the percentage on the vanilla health bar, and something else on the opponent information panel, that's why this toggle exists

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -31,15 +31,25 @@
+import net.runelite.client.callback.ClientThread;
 import net.runelite.client.game.HiscoreManager;
 import net.runelite.client.game.NPCManager;
+
```
Don't add empty lines in the import block.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -60,26 +71,35 @@
 	private int lastRatio = 0;
 	private int lastHealthScale = 0;
 	private String opponentName;
+	private int raidOpponentHealth = 0;
+	private int raidOpponentMaxHealth = 0;
```
int fields default to `0`, you don't need to explicitly set that starting value.
```suggestion
	private int raidOpponentHealth;
	private int raidOpponentMaxHealth;
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -60,26 +71,35 @@
+	private int raidOpponentMaxHealth = 0;
+	private String raidOpponentName;
+	private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
+	boolean useVanilla = false;
+	boolean shouldUnHide = false;
```
boolean fields default to `false`, you don't need to explicitly set that starting value.
```suggestion
	boolean useVanilla;
	boolean shouldUnHide;
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -60,26 +71,35 @@
+	private int raidOpponentHealth = 0;
+	private int raidOpponentMaxHealth = 0;
+	private String raidOpponentName;
+	private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
```
SES should be injected, not instantiated here.

CasvM replied:
With the new code, using varbits instead of the widget, I don't need the scheduler anymore, so I've removed it. Also all the formatting and initialization guidelines you pointed out are added in the latest commit.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -89,44 +109,174 @@ private OpponentInfoOverlay(
+		{
+			clientThread.invokeLater(() ->
+			{
+				Widget raidOpponentWrapper = client.getWidget(303, 6);
+				Widget raidOpponentNameWidget = client.getWidget(303, 11);
+				Widget raidOpponentHealthWidget = client.getWidget(303, 20);
```
These should all be defined in WidgetID and WidgetInfo.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -89,44 +109,174 @@ private OpponentInfoOverlay(
+
+					if (raidOpponentWrapper != null)
+					{
+						//unhide so that initial values can be read again, but wait 7 seconds so that the hp bar isn't shown
+						//while the boss is dying. It's possible to cancel the unhide by changing shouldUnHide back to false
```
What does this mean? Is the varbit for the opponent's health not updated when the widget is hidden?

CasvM replied:
This was the case. When the widget is hidden, the values on the widget don't update. The varbits still do update though, and that is used for the variables in the latest commit.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -89,44 +109,174 @@ private OpponentInfoOverlay(
-		if (opponent.getName() != null && opponent.getHealthScale() > 0)
+		final Actor opponent = opponentInfoPlugin.getLastOpponent();
+
+		if (opponent != null)
```
Why is this guard clause reversed?

CasvM replied:
When the vanilla bar and the opponentinfo are merged, the opponentinfo shouldn't disappear after 5 seconds, or when the opponent is null, since the variables are still known from the vanilla bar. When opponent is null, it shouldn't return yet, but rather go on and check if the raid variables are known. There is an additional gaurd clause which checks if the raid variables are null after this guard clause:
`		if ((opponent == null && (vanillaHealthBarMaxHealth == 0 || !opponentInfoConfig.mergeVanillaHealthBar())) ||
			opponentName == null)
		{
			opponentName = null;
			return null;
		}`

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoOverlay.java:
```diff
@@ -154,8 +304,12 @@ else if (opponent instanceof Player)
-				int health = 0;
-				if (lastRatio > 0)
+
+				//raidOpponentHealth is non-zero if the raid health overlay is active;
+				//The max and min health don't have to be calculated then
+				int health = raidOpponentHealth;
```
What about when the health overlay isn't active?

CasvM replied:
By now I've changed this to `int health = opponentMaxHealth`, so that it only runs when the overlay actually isn't active. Previously it could also run when the opponent simply had zero health. 

But to answer your question: Both the opponenthealth and raidopponenthealth are zero when the overlay isn't active.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoPlugin.java:
```diff
@@ -26,25 +26,20 @@
-import net.runelite.api.events.GameTick;
-import net.runelite.api.events.InteractingChanged;
-import net.runelite.api.events.MenuEntryAdded;
+import net.runelite.api.*;
+import net.runelite.api.events.*;
```
Reverse these wildcard imports.

CasvM replied:
Done in the latest commit. Apologies, I think intellij snuck that in

abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoPlugin.java:
```diff
@@ -164,6 +165,43 @@ public void onGameTick(GameTick gameTick)
+	public void onVarbitChanged(VarbitChanged event)
+	{
+		if ((client.getVar(Varbits.RAID_HEALTH_BAR_OVERLAY_VALUE) != lastHealth) || update ||
+			(config.showVanillaPercentages() && (event.getIndex() == 1682 || event.getIndex() == 1683)))
```
VarbitChanged::getIndex is not a varbit id; this is probably wrong

CasvM replied:
Currently it isn't checking values at all anymore, I could make it like before where it checks if one of the values is changed, but I don't know if that's wanted

abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoPlugin.java:
```diff
@@ -164,6 +165,43 @@ public void onGameTick(GameTick gameTick)
+	@Subscribe
+	public void onVarbitChanged(VarbitChanged event)
+	{
+		if ((client.getVar(Varbits.RAID_HEALTH_BAR_OVERLAY_VALUE) != lastHealth) || update ||
```
VarbitChanged::getIndex is not a varbit id; changing this would be wrong

CasvM commented:
#### Notable changes in the latest commit:

- The 'merge vanilla health bar' feature also works outside of COX.

- The opponent information overlay now takes on the same colours as the boss health bar overlay:
![image](https://user-images.githubusercontent.com/20537001/95632034-125dd700-0a85-11eb-825b-aea40f716dd1.png)
![image](https://user-images.githubusercontent.com/20537001/95632087-2c97b500-0a85-11eb-9708-03f8900041ec.png)
![image](https://user-images.githubusercontent.com/20537001/95632204-6ec0f680-0a85-11eb-9d60-00dfbf22c1cd.png)

Nightfirecat commented:
Aside from code review, I'm not sure what the general opinion is around changing the healthbar colors. I'd think it would be better to not change them since, with the vanilla bar hidden, it would seemingly change for no reason.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/opponentinfo/OpponentInfoPlugin.java:
```diff
@@ -164,6 +190,62 @@ public void onGameTick(GameTick gameTick)
+					clientThread.invokeLater(() ->	opponentInfoOverlay.unHideVanillaHealthBar());
+				}
+			}
+			//reset/check values after config change.
+			clientThread.invokeLater(() ->
+			{
+				opponentInfoOverlay.setVanillaHealthText();
+				opponentInfoOverlay.setHealthBarColors();
+			});
+
+			updateVanillaHealthBarVars();
```
Should these all be running even if mergeVanillaHealthBar wasn't the changed config within opponentinfo?

ulises-c commented:
This is just what I was looking for

destruct7780 commented:
Can this be implemented already? it's crazy we still can't have exact hp values on the runelite hp bar i can't stand the official one
